### PR TITLE
enhm/pytest-requirements

### PR DESCRIPTION
**Pull request by bengsparks**
Collection of small tweaks needed to get more projects up and running, as well as QoL improvements, such as: 

- supporting `requirements.txt` and  `requirements-dev.txt`
- shallow git clones that use SSH
- avoiding overwriting trace data of tests using `@pytest.parametrize`
- logging tracebacks of failed executions to `.err` files
- preferring to hint with None instead of NoneType

### CODE REVIEW

File: tests/typegen/strats/test_inline.py [hunk 1]
```diff
@@ -700,7 +700,7 @@ def function(parameter: Clazz):
     def __init__(self):
         self.class_member: int = 5
 
-    def change_value(self, parameter1: int, parameter2: str) -> None:
+    def change_value(self, parameter1: int, parameter2: str) -> bool:
```
Comment (RecurvedBow) [inline tests/typegen/strats/test_inline.py:703]: It seems like you have found a bug, so I added bug issue #159. You do not need to review this comment.

File: tracing/decorators.py [hunk 2]
```diff
@@ -100,18 +95,44 @@ def _execute_tracing(
             apply_opts=True,
         )
 
-        _trace_callable(tracer, lambda: c(*args, **kwargs))
+        err = _trace_callable(tracer, lambda: c(*args, **kwargs))
 
         traced = tracer.trace_data
 
     if benchmarks is not None:
```
Comment (RecurvedBow) [inline tracing/decorators.py:102]: There is already an if benchmark.performance. Move the code 103 - 115 to the previous if.

File: tracing/decorators.py [hunk 3]
```diff
@@ -100,18 +95,44 @@ def _execute_tracing(
...
+    # Append hash to avoid overwriting other pickled DataFrames
+    trace_subst = config.pytypes.output_template.format_map(
+        {
+            "project": subst.project,
+            "test_case": subst.test_case,
+            "func_name": f"{subst.func_name}-{hash_pandas_object(traced).sum()}",
+        }
+    )
+
     trace_output_path = config.pytypes.proj_path / trace_subst
```
Comment (RecurvedBow) [inline tracing/decorators.py:126]: Move the code from 126 - 135 to line 101, as there is where err is defined.

File: tracing/decorators.py [hunk 4]
```diff
@@ -100,18 +95,44 @@ def _execute_tracing(
...
     trace_output_path.parent.mkdir(parents=True, exist_ok=True)
-    traced.to_pickle(str(trace_output_path))
+
+    if err is not None:
+        err_output_path = trace_output_path.with_suffix(".err")
+        err_output_path.open("w").write(err)
+
+    else:
+        print(trace_output_path)
```
Comment (RecurvedBow) [inline tracing/decorators.py:134]: Remove this line.

File: tracing/decorators.py [hunk 5]
```diff
@@ -90,6 +82,9 @@ def _execute_tracing(
 
         traced = tracers[-1].trace_data
 
+        # Unable to catch error in benchmarking mode due to timeit usage
+        err = None
```
Comment (RecurvedBow) [inline tracing/decorators.py:86]: Drop these lines.

**Review by RecurvedBow — CHANGES_REQUESTED** [review]

File: tracing/decorators.py (hunk 4 above)
Comment (bengsparks) [reply tracing/decorators.py:134]: Done

**Review by RecurvedBow — APPROVED** [review]

**Pull request merged**